Fix order of accounts in `eth_accounts` response

Gudahtt opened the pull request:
The accounts returned by `eth_accounts` were in a fixed order - the order in which the keyring returned them - rather than ordered with the selected account first. The accounts returned by the `accountsChanged` event were ordered with the selected account first, but the same order wasn't used for `eth_accounts`.

We needed to store additional state in order to determine the correct account order correctly on all dapps. We had only been storing the current selected account, but since we also need to determine the primary account per dapp (i.e. the last "selected" account among the accounts exposed to that dapp), that wasn't enough.

A `lastSelected` property has been added to each identity in the preferences controller to keep track of the last selected time. This property is set to the current time (in milliseconds) whenever a new selection is made. The accounts returned with `accountsChanged` and by `eth_accounts` are both ordered by this property.

The `updatePermittedAccounts` function was merged with the internal methods for responding to account selection, to keep things simpler. It wasn't called externally anyway, so it wasn't needed in the public API.

Gudahtt committed:
Fix order of accounts in `eth_accounts` response

The accounts returned by `eth_accounts` were in a fixed order - the
order in which the keyring returned them - rather than ordered with the
selected account first. The accounts returned by the `accountsChanged`
event were ordered with the selected account first, but the same order
wasn't used for `eth_accounts`.

We needed to store additional state in order to determine the correct
account order correctly on all dapps. We had only been storing the
current selected account, but since we also need to determine the
primary account per dapp (i.e. the last "selected" account among the
accounts exposed to that dapp), that wasn't enough.

A `lastSelected` property has been added to each identity in the
preferences controller to keep track of the last selected time. This
property is set to the current time (in milliseconds) whenever a new
selection is made. The accounts returned with `accountsChanged` and by
`eth_accounts` are both ordered by this property.

The `updatePermittedAccounts` function was merged with the internal
methods for responding to account selection, to keep things simpler. It
wasn't called externally anyway, so it wasn't needed in the public API.

Gudahtt commented:
This is a draft because it depends upon #8341

Gudahtt commented on app/scripts/controllers/permissions/index.js:
```diff
@@ -461,46 +438,46 @@ export class PermissionsController {
 
-    const newPermittedAccounts = [account].concat(
-      permittedAccounts.filter((_account) => _account !== account)
+    this.permissions.updateCaveatFor(
```
Do we need to update the caveat if only the order changes? 🤔 
cc @rekmarks 
We had done so before, but I'm not aware that the order of the caveat is meaningful. It just happens to have the same contents as the `eth_accounts` array, where the order _is_ meaningful.

I'm asking because in a later PR I'm adding a `addPermittedAccount` method that adds an account, but it's not necessarily the selected account so I can't easily tell where it belongs in this array.

rekmarks replied:
The order is completely irrelevant as far as `rpc-cap` and the `PermissionsController` are concerned.

Order only matters for account arrays sent to dapps in response to `eth_accounts` requests or for `accountsChanged` events.

Gudahtt replied:
Great! I'll remove it for now then.

Gudahtt replied:
It has been removed.

Gudahtt replied:
This led me to also remove some logic to prevent redundant events: https://github.com/MetaMask/metamask-extension/pull/8342/commits/f90e4474fe0c6319c00137f3ec7427c355c80520

I couldn't think of a way to reliably filter these events without keeping even more state in the controller, so... this seems fine for now, since the provider is already filtering these.

rekmarks replied:
Basically, we're now only sorting the accounts in the `eth_accounts` restricted method implementation, which is called for RPC requests for that method and `accountsChanged` events? You love to see it!

rekmarks replied:
Hm, regarding the complete removal of `updatePermittedAccounts`, are we still allowing the user to "disconnect" individual accounts from websites? If so, how are we handling that?

Gudahtt replied:
That has not yet been implemented. We might resurrect `updatePermittedAccounts` for that, but likely we'd add a `removePermittedAccount` instead, or something like that.

rekmarks replied:
Very good - more attenuated is better. If we add that feature it'll be easy enough to re-implement.

Gudahtt committed:
Remove caveat update upon change in selected account

The order of accounts in the caveat isn't meaningful, so the caveat
doesn't need to be updated when the accounts get re-ordered.

Gudahtt committed:
Emit event regardless of account order

Now that we're no longer relying upon the caveat for the account order,
we also have no way of knowing if a particular account selection
resulted in a change in order or not. The notification is now emitted
whenever an exposed account is selected - even if the order stayed the
same.

The inpage provider currently caches the account order, so it can be
relied upon to ignore these redundant events. We were already emiting
redundant `accountsChanged` events in some cases anyway.

Gudahtt committed:
Remove `getIdentities` getter

This getter was no longer required, since the entire preferences store
is now being passed to the permissions controller.

rekmarks commented on app/scripts/controllers/permissions/restrictedMethods.js:
```diff
@@ -1,12 +1,28 @@
           .then((accounts) => {
+            const identities = getIdentities()
             res.result = accounts
+              .sort((firstAddress, secondAddress) => {
+                if (!identities[firstAddress]) {
+                  throw new Error(`Missing identity for address ${firstAddress}`)
+                } else if (!identities[secondAddress]) {
+                  throw new Error(`Missing identity for address ${secondAddress}`)
+                } else if (identities[firstAddress].lastSelected === identities[secondAddress].lastSelected) {
+                  return 0
+                } else if (identities[firstAddress].lastSelected === undefined) {
+                  return 1
+                } else if (identities[secondAddress].lastSelected === undefined) {
+                  return -1
+                }
+
+                return identities[secondAddress].lastSelected - identities[firstAddress].lastSelected
+              })
```
Since index 0 of the accounts array is the only one with any meaning, do we need to sort the whole thing?

The alternative would be to find the `Math.min` of the identities by their `lastSelected` property, put the corresponding account first, if any, and concatenate the rest.

Gudahtt replied:
>Ah, no we can't, because the addresses are filtered by the caveats after this call.

Yep, exactly. So because of how the caveat filter works, we're kinda forced to order them by last-selected time.

rekmarks replied:
Also, several dozen things will break before the sorting becomes a performance problem. This is definitely fine for now. It'd be nice if this sorting didn't have to happen every call to `eth_accounts`, but we can figure that out later.

rekmarks reviewed (approved):
LGTM!
